Jasmine/Tetris/ViewModels/TetrisGameViewModel.swift

```diff
@@ -1,4 +1,200 @@
+
+    private(set) var gameStatus = GameStatus.notStarted {
+        didSet {
+            delegate?.notifyGameStatusUpdated()
+        }
+    }
+
+    /// Populate upcomingTiles
+    init() {
+        populateUpcomingTiles()
+        setTimerListener()
+    }
+
+    private func populateUpcomingTiles() {
+        for _ in 0..<Constants.Game.Tetris.upcomingTilesCount {
+            upcomingTiles.append(getRandomWord())
+        }
+    }
+
+    private func setTimerListener() {
+        timer.timerListener = { status in
+            switch status {
+            case .start:
+                self.gameStatus = .inProgress
+                self.delegate?.redisplay(timeRemaining: self.timeRemaining, outOf: self.totalTimeAllowed)
+            case .tick:
+                self.delegate?.redisplay(timeRemaining: self.timeRemaining, outOf: self.totalTimeAllowed)
+            case .finish:
+                self.gameStatus = .endedWithLost
+            case .stop:
+                self.gameStatus = .endedWithWon
+            }
+        }
+    }
+
+    /// Checks for and returns coordinates of matching phrase, searching by row-wise and return if found
+    /// otherwise continue with searching by column-wise
+    fileprivate func checkForMatchingPhrase() -> Set<Coordinate>? {
+        return checkForMatchingPhrase(byRow: true) ?? checkForMatchingPhrase(byRow: false)
+    }
+
+    /// Checks for and returns coordinates of matching phrase, search row/col-wise as specify by searchbyRow
+    /// Concatenate the words row by row or col by col to check if a phrase is contained in them
+    private func checkForMatchingPhrase(byRow searchByRow: Bool) -> Set<Coordinate>? {
+        var matchedCoordinates: Set<Coordinate> = []
+        let maxIndex = searchByRow ? Constants.Game.Tetris.rows : Constants.Game.Tetris.columns
+        for index in 0..<maxIndex {
+            var line = ""
+            if searchByRow {
+                for col in 0..<Constants.Game.Tetris.columns {
+                    line += getTileText(at: Coordinate(row: index, col: col))
+                }
+            } else {
+                for row in 0..<Constants.Game.Tetris.rows {
+                    line += getTileText(at: Coordinate(row: row, col: index))
+                }
+            }
+
+            guard let validPhraseRange = getValidPhraseRange(line) else {
+                continue
+            }
+
+            for i in validPhraseRange {
+                let coordinate = searchByRow ? Coordinate(row: index, col: i) : Coordinate(row: i, col: index)
+                matchedCoordinates.insert(coordinate)
+            }
+            return matchedCoordinates
+        }
+        return nil
+    }
+
+    /// Gets the tile text at coordinate specified by `row` and `col`
+    /// returns " " if no tile is present so that phrases separated by gaps don't get matched
+    private func getTileText(at coordinate: Coordinate) -> String {
```

Comment:
I would make this return `String?`. What if suddenly we need to match those that has a blank space too :joy: